File: packages/client-java/src/main/java/com/durablestreams/internal/sse/SSEParser.java
package com.durablestreams.internal.sse;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.*;

/**
 * Parser for Server-Sent Events (SSE) streams.
 */
public final class SSEParser {

    private final BufferedReader reader;
    private boolean closed;

    public SSEParser(InputStream input) {
        this.reader = new BufferedReader(new InputStreamReader(input, StandardCharsets.UTF_8));
        this.closed = false;
    }

    /**
     * Read the next SSE event.
     * Returns null if the stream is closed.
     */
    public SSEEvent nextEvent() throws IOException {
        if (closed) return null;

        StringBuilder data = new StringBuilder();
        String eventType = "message";
        String id = null;
        Integer retry = null;

        String line;
        while ((line = reader.readLine()) != null) {
            if (line.isEmpty()) {
                // Empty line marks end of event
                if (data.length() > 0) {
                    // Remove trailing newline if present
                    if (data.charAt(data.length() - 1) == '\n') {
                        data.setLength(data.length() - 1);
                    }
                    return new SSEEvent(eventType, data.toString(), id, retry);
                }
                continue;
            }

            if (line.startsWith(":")) {
                // Comment, ignore
                continue;
            }

            int colonIndex = line.indexOf(':');
            String field, value;
            if (colonIndex == -1) {
                field = line;
                value = "";
            } else {
                field = line.substring(0, colonIndex);
                value = line.substring(colonIndex + 1);
                // Remove leading space from value if present
                if (value.startsWith(" ")) {
                    value = value.substring(1);
                }
            }

            switch (field) {
                case "event":
                    eventType = value;
                    break;
                case "data":
                    data.append(value).append("\n");
                    break;
                case "id":
                    id = value;
                    break;
                case "retry":
                    try {
                        retry = Integer.parseInt(value);
                    } catch (NumberFormatException e) {
                        // Ignore invalid retry values
                    }
                    break;
                default:
                    // Unknown field, ignore
                    break;
            }
        }

        // End of stream
        closed = true;
        if (data.length() > 0) {
            if (data.charAt(data.length() - 1) == '\n') {
                data.setLength(data.length() - 1);
            }
            return new SSEEvent(eventType, data.toString(), id, retry);
        }
        return null;
    }

    /**
     * Check if the parser is closed.
     */
    public boolean isClosed() {
        return closed;
    }

    /**
     * Close the parser and underlying stream.
     */
    public void close() throws IOException {
        closed = true;
        reader.close();
    }

    /**
     * Represents a single SSE event.
     */
    public static final class SSEEvent {
        private final String event;
        private final String data;
        private final String id;
        private final Integer retry;

        public SSEEvent(String event, String data, String id, Integer retry) {
            this.event = event;
            this.data = data;
            this.id = id;
            this.retry = retry;
        }

        public String getEvent() {
            return event;
        }

        public String getData() {
            return data;
        }

        public Optional<String> getId() {
            return Optional.ofNullable(id);
        }

        public Optional<Integer> getRetry() {
            return Optional.ofNullable(retry);
        }

        /**
         * Check if this is a control event (contains stream metadata).
         */
        public boolean isControl() {
            return "control".equals(event);
        }

        @Override
        public String toString() {
            return "SSEEvent{" +
                   "event='" + event + '\'' +
                   ", data='" + data + '\'' +
                   ", id='" + id + '\'' +
                   ", retry=" + retry +
                   '}';
        }
    }
}

File: packages/client-dotnet/src/DurableStreams/Sse/SseParser.cs
using System.Buffers;
using System.Text;
using System.Text.Json;

namespace DurableStreams.Sse;

/// <summary>
/// SSE event types.
/// </summary>
internal enum SseEventType
{
    Data,
    Control
}

/// <summary>
/// Parsed SSE data event.
/// </summary>
internal readonly record struct SseDataEvent(string Data);

/// <summary>
/// Parsed SSE control event.
/// </summary>
internal readonly record struct SseControlEvent(
    string StreamNextOffset,
    string? StreamCursor,
    bool UpToDate,
    bool StreamClosed);

/// <summary>
/// Parser for Server-Sent Events stream.
/// </summary>
internal sealed class SseParser : IDisposable
{
    private readonly StreamReader _reader;
    private readonly StringBuilder _dataBuffer = new();
    private string? _eventType;
    private bool _disposed;

    public SseParser(Stream stream)
    {
        _reader = new StreamReader(stream, Encoding.UTF8, detectEncodingFromByteOrderMarks: false);
    }

    /// <summary>
    /// Read the next SSE event.
    /// Returns null on end of stream.
    /// </summary>
    public async Task<(SseEventType Type, object Event)?> ReadEventAsync(CancellationToken cancellationToken = default)
    {
        while (!_disposed)
        {
            var line = await _reader.ReadLineAsync(cancellationToken).ConfigureAwait(false);

            if (line == null)
            {
                // End of stream - try to flush any pending event
                var finalEvent = FlushEvent();
                return finalEvent;
            }

            // Empty line signals end of event
            if (string.IsNullOrEmpty(line))
            {
                var evt = FlushEvent();
                if (evt != null)
                {
                    return evt;
                }
                continue;
            }

            // Parse line
            if (line.StartsWith("event:", StringComparison.Ordinal))
            {
                _eventType = line[6..].TrimStart();
            }
            else if (line.StartsWith("data:", StringComparison.Ordinal))
            {
                var content = line[5..];
                // Strip optional leading space
                if (content.StartsWith(' '))
                {
                    content = content[1..];
                }

                if (_dataBuffer.Length > 0)
                {
                    _dataBuffer.Append('\n');
                }
                _dataBuffer.Append(content);
            }
            // Ignore id:, retry:, and comment lines (:)
        }

        return null;
    }

    private (SseEventType Type, object Event)? FlushEvent()
    {
        // For data events, require non-empty data
        // For control events, still try to parse even if empty (empty is invalid JSON)
        if (_eventType == null)
        {
            _eventType = null;
            _dataBuffer.Clear();
            return null;
        }

        // Skip data events with empty content
        if (_eventType == "data" && _dataBuffer.Length == 0)
        {
            _eventType = null;
            _dataBuffer.Clear();
            return null;
        }

        var data = _dataBuffer.ToString();
        var eventType = _eventType;

        _eventType = null;
        _dataBuffer.Clear();

        return eventType switch
        {
            "data" => (SseEventType.Data, new SseDataEvent(data)),
            "control" => ParseControlEvent(data),
            _ => HandleUnknownEvent(eventType, data)
        };
    }

    private static (SseEventType Type, object Event)? HandleUnknownEvent(string eventType, string data)
    {
        // Log unknown event types for debugging protocol mismatches
        System.Diagnostics.Debug.WriteLine(
            $"[DurableStreams] Unknown SSE event type '{eventType}'. " +
            $"This may indicate a protocol version mismatch. Data: {data.Substring(0, Math.Min(100, data.Length))}");
        return null;
    }

    private static (SseEventType Type, object Event)? ParseControlEvent(string data)
    {
        try
        {
            using var doc = JsonDocument.Parse(data);
            var root = doc.RootElement;

            var streamNextOffset = root.TryGetProperty("streamNextOffset", out var offsetProp)
                ? offsetProp.GetString() ?? ""
                : "";

            var streamCursor = root.TryGetProperty("streamCursor", out var cursorProp)
                ? cursorProp.GetString()
                : null;

            var streamClosed = root.TryGetProperty("streamClosed", out var streamClosedProp) &&
                               (streamClosedProp.ValueKind == JsonValueKind.True ||
                                (streamClosedProp.ValueKind == JsonValueKind.String &&
                                 string.Equals(streamClosedProp.GetString(), "true", StringComparison.OrdinalIgnoreCase)));

            var upToDate = root.TryGetProperty("upToDate", out var upToDateProp) &&
                          (upToDateProp.ValueKind == JsonValueKind.True ||
                           (upToDateProp.ValueKind == JsonValueKind.String &&
                            string.Equals(upToDateProp.GetString(), "true", StringComparison.OrdinalIgnoreCase)));

            // Closed streams are implicitly up-to-date
            if (streamClosed)
            {
                upToDate = true;
            }

            return (SseEventType.Control, new SseControlEvent(streamNextOffset, streamCursor, upToDate, streamClosed));
        }
        catch (JsonException ex)
        {
            // Control events contain critical stream position data - don't silently ignore parse failures
            var preview = data.Length > 100 ? data[..100] + "..." : data;
            throw new DurableStreamException(
                $"Failed to parse SSE control event: {ex.Message}. Data: {preview}",
                DurableStreamErrorCode.ParseError,
                null,
                null,
                ex);
        }
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        _reader.Dispose();
    }
}
